kratos/includes/global_pointer.h

```diff
@@ -0,0 +1,186 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Carlos A. Roig
+//
+
+#if !defined(KRATOS_GLOBAL_POINTER_H_INCLUDED )
+#define  KRATOS_GLOBAL_POINTER_H_INCLUDED
+
+#include <iostream>
+
+#include "includes/define.h"
+
+#ifdef KRATOS_USING_MPI
+#include "mpi.h"
+#endif
+
+namespace Kratos {
+
+template<class TDataType>
+class GlobalPointer {
+private:
+
+  TDataType GetPointer() { return mDataPointer; }
+
+  /// Pointer to the data
+  TDataType * mDataPointer;
+
+#ifdef KRATOS_USING_MPI
+  /// Rank is enabled only under mpi
+  int mRank;
+#endif
+
+public:
+
+	/** Default constructor
+	* Default constructor
+	* This should never be called as we need a local pointer to exists
+	*/
+	GlobalPointer() = delete;
+
+  /** Constructor by Data
+   * Constructor by Data
+   * @param Data the data.
+   */
+  GlobalPointer(TDataType Data) = delete;
+
+  /** Constructor by Data Pointer
+   * Constructor by Data Pointer
+   * @param DataPointer Pointer to the data.
+   */
+  GlobalPointer(TDataType * DataPointer)
+    : mDataPointer(DataPointer)
+#ifdef KRATOS_USING_MPI
+    , mRank(GetLocalRank())
+#endif
+    {
+  }
+
+  /** Constructor by boost::shared_ptr
+   * Constructor by boost::shared_ptr
+   * @param DataPointer Boost Shared Pointer to the Data.
+   */
+  GlobalPointer(boost::shared_ptr<TDataType> DataPointer)
+    : mDataPointer(DataPointer.get())
+#ifdef KRATOS_USING_MPI
+    , mRank(GetLocalRank())
+#endif
+    {
+  }
+
+  /** Constructor by boost::weak_ptr
+   * Constructor by boost::weak_ptr
+   * @param DataPointer Boost Weak Pointer to the Data.
+   */
+  GlobalPointer(boost::weak_ptr<TDataType> DataPointer) = delete;
```

Comment:
I suppose no. anyway, its odd since this IMHO can be seen from two different sides:

1 - We use the current implementation of the global_ptr:
It means that the lock needs to be done when acquiring the data, so once the user creates the global_ptr, the old weak_ptr is useless, has he can easily bypass the lock by accessing the data directly from the global_ptr.

2 - We store the weak_ptr inside the global_ptr:
Then the lock can be made inside the pointer, but it needs a completely separate implementation in order to ensure that honors the lock mechanism.

I uploaded following 1, but I would not recommend using this...